presto-main/src/main/java/com/facebook/presto/server/security/LdapFilter.java

```diff
@@ -106,153 +131,171 @@ public void doFilter(ServletRequest servletRequest, ServletResponse servletRespo
-            try {
-                context = (DirContext) authenticationCache.get(environment);
-            }
-            catch (ExecutionException e) {
-                Throwable throwable = e.getCause();
-                if (throwable instanceof AuthenticationException) {
-                    throw new AuthenticationException(throwable.getMessage());
-                }
-                else {
-                    throw Throwables.propagate(e);
-                }
-            }
-
-            log.debug("Authentication successful for user %s.", user);
-
-            if (genericLdapBinder.getGroupAuthorizationSearchPattern().isPresent()) {
-                checkForGroupMembership(genericLdapBinder.getGroupAuthorizationSearchPattern().get(), user, context);
-            }
-
+            String header = request.getHeader(AUTHORIZATION);
+            Credentials credentials = getCredentials(header);
+            Principal principal = getPrincipal(credentials);
             // ldap authentication ok, continue
             nextFilter.doFilter(new HttpServletRequestWrapper(request)
             {
                 @Override
                 public Principal getUserPrincipal()
                 {
-                    return new LdapPrincipal(user);
+                    return principal;
                 }
             }, servletResponse);
         }
         catch (AuthenticationException e) {
-            // log the failure since the exception goes directly to the client
+            log.debug(e, "LDAP authentication failed");
+            if (e.getStatus() == UNAUTHORIZED) {
+                sendChallenge(request, response, e.getMessage());
+                return;
+            }
+            response.sendError(e.getStatus().code(), e.getMessage());
+        }
+    }
+
+    private static void sendChallenge(HttpServletRequest request, HttpServletResponse response, String errorMessage)
+            throws IOException
+    {
+        long contentLength = request.getContentLengthLong();
+        if (contentLength > 0) {
+            skipFully(request.getInputStream(), contentLength);
+        }
+        response.setHeader(HttpHeaders.WWW_AUTHENTICATE, "Basic realm=\"presto\"");
+        response.sendError(UNAUTHORIZED.code(), errorMessage);
+    }
+
+    private static Credentials getCredentials(String header)
+            throws AuthenticationException
+    {
+        if (header == null) {
+            throw new AuthenticationException(UNAUTHORIZED, "Unauthorized");
+        }
+        if (!header.startsWith(BASIC_AUTHENTICATION_PREFIX)) {
+            throw new AuthenticationException(BAD_REQUEST, "Basic authentication is expected");
+        }
+        String base64Credentials = header.substring(BASIC_AUTHENTICATION_PREFIX.length());
+        String credentials = new String(Base64.getDecoder().decode(base64Credentials), UTF_8);
```

Comment:
This version of the code handles an incorrect request differently than before: now, if there were more than two \s-delimited parts, this decode() method will throw an exception and you'll get 500, rather than an "Invalid credentials" error (see line 118 of the old code).